python/plugins/processing/algs/qgis/CreateAtlasGrid.py

```diff
@@ -0,0 +1,197 @@
+        # The name that the user will see in the toolbox
+        return self.tr('Create a grid from map layout')
+
+    def group(self):
+        return self.tr('Cartography')
+
+    def groupId(self):
+        return 'cartography'
+
+    def __init__(self):
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+
+        param = QgsProcessingParameterLayout(self.LAYOUT, 'Print layout')
+        self.addParameter(param)
+        map_param = QgsProcessingParameterLayoutItem(self.MAP, 'Map item', parentLayoutParameterName=self.LAYOUT,
+                                                     itemType=QgsLayoutItemRegistry.LayoutMap)
+        self.addParameter(map_param)
+        self.addParameter(QgsProcessingParameterFeatureSource(self.COVERAGE_LAYER,
+                                                              self.tr('Coverage layer'),
+                                                              [QgsProcessing.TypeMapLayer]))
+        self.addParameter(QgsProcessingParameterDistance(self.COVERAGE_BUFFER,
+                                                         self.tr('Coverage buffer'), parentParameterName='COVERAGE_LAYER',
+                                                         defaultValue=0.0))
+        self.addParameter(QgsProcessingParameterBoolean(self.ONLY_ON_FEATURES,
+                                                        self.tr('Only on intersected features'),
+                                                        defaultValue=True))
+        self.addParameter(QgsProcessingParameterFeatureSink(self.OUTPUT,
+                                                            self.tr('Grid'),
+                                                            QgsProcessing.TypeVectorPolygon))
+
+    def prepareAlgorithm(self, parameters, context, feedback):
+
+        layout = self.parameterAsLayout(parameters, self.LAYOUT, context)
+        if layout is None:
+            raise QgsProcessingException('Cannot find layout with name "{}"'.format(parameters[self.LAYOUT]))
+
+        map = self.parameterAsLayoutItem(parameters, self.MAP, context, layout)
+        if map is None:
+            raise QgsProcessingException('Cannot find matching map item with ID {}'.format(parameters[self.MAP]))
+
+        self.mapheight = map.extent().height()
+        self.mapwidth = map.extent().width()
+
+        self.coverage_layer = self.parameterAsVectorLayer(
+            parameters,
+            self.COVERAGE_LAYER,
+            context
+        )
+        coverage_distance = self.parameterAsDouble(
+            parameters,
+            self.COVERAGE_BUFFER,
+            context
+        )
+        self.only_on_features = self.parameterAsBool(
+            parameters,
+            self.ONLY_ON_FEATURES,
+            context
+        )
+        self.coverage = self.coverage_layer.sourceExtent().buffered(coverage_distance)
+
+        self.total = (self.coverage.height() / self.mapheight) * (self.coverage.width() / self.mapwidth)
```

Comment:
I'm unsure - does this approach work ok with rotated maps?

Reply:
Sorry for the delay. I really don´t take care of rotated maps. In fact, the alg always generate grids as unrotated.

Reply:
IMHO can be merged @nyalldawson but not before to add a warning in alg description that it does not manage rotated maps.

Reply:
Hm, I disagree - we'd be introducing a new feature with a known bug. Ultimately it's adding another bug that becomes the project's responsibility to fix... And we all know how overworked the bug fixing team is currently!

Reply:
ok, I agree with you... @fpsampayo before the requests was implicit now it's ore than explicit, any idea to adapt it with some more cos/sin ;)

Reply:
Ok, I don´t really understand the problem, do you want that if map have rotation, the output grid shall be rotated too? In my mind this not have sense but i can try to implement it.

Reply:
Yes - if the map item content is rotated, then the grid should also be rotated so that individual grid feature polygons exactly match the map rotation. Otherwise it's not possible to use this to create an atlas grid linked to a rotated map

Reply:
Ok @nyalldawson, i´ll work on this but i don´t have time just now and I need, I think, to rewrite the code. Because of this, maybe, if you consider, we can close this PR and create new one when I done. 
And I want to thank you all the reviews and tips to be able to arrive at this point.